text/0000-overlapping_match.md

```diff
@@ -0,0 +1,183 @@
+- Feature Name: overlapping_match_statements
+- Start Date: 2017-09-08
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This idea facilitates the writing and using of `match` expressions where multiple branches are
+executed. Writing `match` expressions with this idea allows for multiple branches to be matched
+and for a check on no matches as well, similar to the current use of the `_` pattern.
+
+# Motivation
+[motivation]: #motivation
+
+There is a very good software engineering principle where repeating a piece of code is bad.
+This is the case because if that selection of code needs to be changed then it has to be
+changed in two places which can easily not be done and thus create bugs. A way of doing this
+for a large selection of lines of code is to put it into a function, a helper function. Allowing
+overlapping match statements extends this paradigm to that where matching is a good idea, the
+use of pattern matching, and where exhaustiveness checks are a nice thing.
+
+This would support use cases where the required execution of several branches overlapped enough
+that his would help. A use case for this is when the outcome of one branch is the same as a
+combination of the other two branches of a match statement. The expected outcome of this is
+the ability to have multiple branches of a match statement, and having those branches still be
+checked for exhaustiveness, be executed if more than one of them match the value.
+
+# Detailed design
+[design]: #detailed-design
+
+Basic Syntax:
+```rust
+match val in {
+    pat | pat => expr,
+    pat => expr
+}
+
+match val in {
+    pat | pat => expr,
+    pat => expr
+} else {
+    expr
+}
+```
+
+Benefits of this syntax:
+1. No new keywords need to be used. This is good thing since it means for a relatively small
+addition there would be no code breaks of existing code with this change.
+2. The `in` seems to imply that it sort of like an "iterator" of statements and will go through
+each of them in turn.
+
```

Comment:
1. Semantically, I consider this a big addition - which is why I'd like a more visible syntactic difference where this RFC to be accepted. Therefore, I consider the slight variation in syntax to be a downside. I expand on this further down.

Instead of `in`, I'd recommend using a terminal that more clearly indicates that this is top-down and more crucially that multiple branches may be taken. Perhaps the syntax might be: `match inclusive <expr> { <patterns> }`, or `match topdown <expr> { <patterns> }`.

2. I don't agree that `in` implies this nearly strong enough for this to be instantly understandable.

Reply:
That is fair and I am pretty sure that adding an additional keyword here wouldn't break any code since if a variable was called what ever it was that would be able to be checked but that is definitely something to consider

Reply:
Yes, adding the additional "keyword" should work w/o backwards incompatibility since there's no (iirc) parsing rule "<expr> <expr>".

Reply:
For a correct DFA adding that state would be rather complicated but I agree is do able

Reply:
Hmm... DFA for what? Lexing? The language is not regular...
Do note that parsing `Rust` is already at least context sensitive.

Reply:
I have not actually looked that deep into Rust I didn't consider it to be that high order, though tbh I don't know why

Reply:
Rust's lexical grammar is already context-sensitive due to `r##"strings like this"##`. The actual grammar after tokenization makes do with a fixed lookahead (I forgot how much exactly, something in the order of 5 tokens or so).

Not a single real programming language is regular (not counting assembly and machine code). Fun fact: Even the simple-looking Lua grammar requires infinite lookahead to parse and is thus harder to parse than Rust.

Reply:
More details here: https://stackoverflow.com/a/43693150/1063961 and here https://stackoverflow.com/a/43693194/1063961